File: day10/day10.go
package main

import (
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func main() {
	data, err := io.ReadAll(os.Stdin)
	if err != nil {
		fmt.Println("error reading from stdin:", err)
		os.Exit(1)
	}

	dataTrimmed := strings.TrimRight(string(data), "\n")
	lines := strings.Split(string(dataTrimmed), "\n")

	var sol1 int
	for _, line := range lines {
		wantState, buttons, _ := parse(line)

		initState := make(State, len(wantState))
		states := []State{initState}
		for {
			sol1++
			var updatedStates []State
			for _, b := range buttons {
				for _, s := range states {
					update := s.Apply(b)
					updatedStates = append(updatedStates, update)
					if update.Equal(wantState) {
						goto done
					}
				}
			}
			states = updatedStates
		}
	done:
	}

	// Part 2 is done in Python
	// There is no good lib for LP problems
	// (◡︵◡)

	fmt.Println("sol1:", sol1)
}

type Joltage []int
type Button []int
type State []bool

func (s State) Copy() State {
	cpy := make(State, len(s))
	copy(cpy, s)
	return cpy
}

func (s State) Apply(b Button) State {
	newState := s.Copy()
	for _, i := range b {
		newState[i] = !newState[i]
	}
	return newState
}

func (s State) Equal(other State) bool {
	if len(s) != len(other) {
		return false
	}
	for i := range s {
		if s[i] != other[i] {
			return false
		}
	}
	return true
}

// parse parses line to desired states, buttons, joltage
func parse(line string) (State, []Button, Joltage) {
	var states State
	var buttons []Button
	var joltage Joltage

	for seg := range strings.SplitSeq(line, " ") {
		switch seg[0] {
		case '[':
			for _, r := range seg[1 : len(seg)-1] {
				on := r == '#'
				states = append(states, on)
			}
		case '(':
			var button []int
			for numString := range strings.SplitSeq(seg[1:len(seg)-1], ",") {
				num, err := strconv.Atoi(numString)
				if err != nil {
					panic(err)
				}
				button = append(button, num)
			}
			buttons = append(buttons, button)
		case '{':
			for numString := range strings.SplitSeq(seg[1:len(seg)-1], ",") {
				num, err := strconv.Atoi(numString)
				if err != nil {
					panic(err)
				}
				joltage = append(joltage, num)
			}
		default:
			fmt.Println("unknown seg start:", seg)
			panic("failed to parse line")
		}
	}

	return states, buttons, joltage
}

File: day10/day10.py
# /// script
# dependencies = [
#   "scipy",
#   "numpy",
# ]
# ///

# Usage:
# cat in.txt | uv run day10.py

import sys
import numpy as np
from scipy.optimize import linprog


def main():
    try:
        data = sys.stdin.read()
    except EOFError:
        return

    if not data:
        return

    data_trimmed = data.rstrip("\n")
    lines = data_trimmed.split("\n")

    sol2 = 0
    for line in lines:
        if not line.strip():
            continue

        _, buttons, want_joltage = parse(line)

        c = [1 for _ in range(len(buttons))]
        b_eq = np.array(want_joltage)
        A_eq = np.array(
            [[int(j in button) for button in buttons] for j in range(len(want_joltage))]
        )

        res = linprog(c=c, A_eq=A_eq, b_eq=b_eq, integrality=1, method="highs")

        if res.success:
            sol2 += int(round(res.fun))

    print(f"sol2: {sol2}")


def parse(line):
    states = []
    buttons = []
    joltage = []

    segments = line.split(" ")
    for seg in segments:
        if not seg:
            continue

        if seg.startswith("["):
            for char in seg[1:-1]:
                states.append(char == "#")

        elif seg.startswith("("):
            nums = [int(n) for n in seg[1:-1].split(",") if n]
            buttons.append(nums)

        elif seg.startswith("{"):
            nums = [int(n) for n in seg[1:-1].split(",") if n]
            joltage.extend(nums)

        else:
            print(f"unknown seg start: {seg}")
            raise ValueError("failed to parse line")

    return states, buttons, joltage


if __name__ == "__main__":
    main()
